File: server-go/state.go
package server

import (
	"sync"
)

// PasskeyAnswer mirrors state.rs::PasskeyAnswer.
type PasskeyAnswer struct {
	// Assertion is set when the app returned a WebAuthn assertion.
	Assertion *PasskeyAssertion
	// Cancelled is set when the user cancelled.
	Cancelled bool
}

// PasskeyAssertion carries the decoded WebAuthn assertion from the app.
type PasskeyAssertion struct {
	CredentialID []byte
	AssertionJSON []byte
}

// AppState mirrors state.rs::AppState.
type AppState struct {
	Cfg   *Config
	Store *HistoryStore
	Bus   *EventBus

	// Current QR string (while pairing).
	qrMu   sync.Mutex
	qr     string
	// Linked phone number.
	phoneMu sync.Mutex
	phone   string
	// 8-char pair code issued by /pair-code.
	pairCodeMu sync.Mutex
	pairCode   string
	// Last passkey request options JSON (recoverable via /passkey/request).
	passkeyPendingMu sync.Mutex
	passkeyPending   string
	// Pending passkey ceremony answer (the app resolves it via
	// /passkey/assertion or /passkey/cancel).
	passkeyAnswerMu sync.Mutex
	passkeyAnswer   chan PasskeyAnswer
	// Call registry (populated by calls.go).
	Calls *CallRegistry
	// Media cache (downloads incoming media; serves /media/{id}/{kind}).
	Media *mediaCache
	// Avatar cache (serves /avatar?chat=...).
	Avatars *avatarCache
	// Followed-channel metadata cache (serves /channels).
	ChannelCache *channelCache
}

// NewAppState builds the shared state.
func NewAppState(cfg *Config) (*AppState, error) {
	store, err := NewHistoryStore(cfg)
	if err != nil {
		return nil, err
	}
	return &AppState{
		Cfg:          cfg,
		Store:        store,
		Bus:          NewEventBus(),
		Calls:        NewCallRegistry(),
		Media:        newMediaCache(cfg.MediaDir),
		Avatars:      newAvatarCache(cfg.DBPath),
		ChannelCache: &channelCache{},
	}, nil
}

// IsLinked reports whether a phone number is set.
func (s *AppState) IsLinked() bool {
	s.phoneMu.Lock()
	defer s.phoneMu.Unlock()
	return s.phone != ""
}

// Phone returns the linked phone number.
func (s *AppState) Phone() string {
	s.phoneMu.Lock()
	defer s.phoneMu.Unlock()
	return s.phone
}

// SetPhone records the linked phone number.
func (s *AppState) SetPhone(phone string) {
	s.phoneMu.Lock()
	s.phone = phone
	s.phoneMu.Unlock()
	s.Store.SetPhone(phone)
}

// QR returns the current QR string.
func (s *AppState) QR() string {
	s.qrMu.Lock()
	defer s.qrMu.Unlock()
	return s.qr
}

// SetQR updates the current QR string.
func (s *AppState) SetQR(qr string) {
	s.qrMu.Lock()
	s.qr = qr
	s.qrMu.Unlock()
}

// PairCode returns the last-issued pair code.
func (s *AppState) PairCode() string {
	s.pairCodeMu.Lock()
	defer s.pairCodeMu.Unlock()
	return s.pairCode
}

// SetPairCode records the last-issued pair code.
func (s *AppState) SetPairCode(code string) {
	s.pairCodeMu.Lock()
	s.pairCode = code
	s.pairCodeMu.Unlock()
}

// PasskeyPending returns and clears the pending request options.
func (s *AppState) TakePasskeyPending() (string, bool) {
	s.passkeyPendingMu.Lock()
	defer s.passkeyPendingMu.Unlock()
	if s.passkeyPending == "" {
		return "", false
	}
	opts := s.passkeyPending
	s.passkeyPending = ""
	return opts, true
}

// SetPasskeyPending stores the latest request options.
func (s *AppState) SetPasskeyPending(options string) {
	s.passkeyPendingMu.Lock()
	s.passkeyPending = options
	s.passkeyPendingMu.Unlock()
}

// InstallPasskeyAnswer replaces (cancelling prior) the pending ceremony answer
// channel. Returns the channel the app's /passkey/assertion or /passkey/cancel
// resolves.
func (s *AppState) InstallPasskeyAnswer() chan PasskeyAnswer {
	s.passkeyAnswerMu.Lock()
	defer s.passkeyAnswerMu.Unlock()
	if old := s.passkeyAnswer; old != nil {
		select {
		case old <- PasskeyAnswer{Cancelled: true}:
		default:
		}
	}
	ch := make(chan PasskeyAnswer, 1)
	s.passkeyAnswer = ch
	return ch
}

// TakePasskeyAnswer removes the pending answer channel (the ceremony ended).
func (s *AppState) TakePasskeyAnswer() chan PasskeyAnswer {
	s.passkeyAnswerMu.Lock()
	defer s.passkeyAnswerMu.Unlock()
	ch := s.passkeyAnswer
	s.passkeyAnswer = nil
	return ch
}

File: server/src/state.rs
//! Shared state: config, the in-memory history store, and the event broadcast
//! channel that the WebSocket server subscribes to.

use anyhow::Result;
use std::sync::Arc;
use tokio::sync::broadcast;

use crate::events::Config;
use crate::store::HistoryStore;

/// What the app sends back to resolve a pending passkey (WebAuthn) request.
#[derive(Debug)]
pub enum PasskeyAnswer {
    /// The WebAuthn assertion, ready for the `<passkey_prologue>` IQ.
    Assertion(whatsapp_rust::passkey::Assertion),
    /// The user cancelled the passkey ceremony.
    Cancelled,
}

/// Outbound events pushed to app WebSocket clients.
#[derive(Clone, Debug, serde::Serialize)]
pub enum Event {
    #[serde(rename = "qr")]
    Qr { qr: String },
    #[serde(rename = "linked")]
    Linked { phone: String },
    #[serde(rename = "message")]
    Message { chat: String },
    #[serde(rename = "receipt")]
    Receipt { chat: String },
    #[serde(rename = "chats")]
    Chats,
    #[serde(rename = "logged_out")]
    LoggedOut,
    #[serde(rename = "incoming_call")]
    IncomingCall { call_id: String, from: String, video: bool },
    #[serde(rename = "call_state")]
    CallState { call_id: String, state: String },
    #[serde(rename = "pair_code")]
    PairCode { code: String },
    /// WhatsApp's SHORTCAKE passkey gate wants a WebAuthn assertion. The app
    /// must run `navigator.credentials.get`-style flow and POST it back.
    #[serde(rename = "passkey_request")]
    PasskeyRequest { options: String },
    /// A verification code to show on the phone (fresh-link confirmation).
    #[serde(rename = "passkey_confirmation")]
    PasskeyConfirmation { code: String },
    #[serde(rename = "passkey_error")]
    PasskeyError { error: String },
}

/// The server's mutable state, shared across the bot callbacks and HTTP handlers.
#[derive(Clone)]
pub struct AppState {
    pub cfg: Arc<Config>,
    pub store: Arc<HistoryStore>,
    /// Broadcast channel for WS events. Subscribers receive every event.
    pub tx: broadcast::Sender<Event>,
    /// Current QR code (while pairing) — read by `/status`.
    pub current_qr: Arc<std::sync::Mutex<String>>,
    /// Phone number once linked.
    pub phone: Arc<std::sync::Mutex<String>>,
    /// Active WhatsApp calls (media relay handles).
    pub calls: crate::calls::CallRegistry,
    /// The 8-char pair code issued by the phone-number flow ("" when none).
    pub pair_code: Arc<std::sync::Mutex<String>>,
    /// Last passkey request options JSON (so the app can recover if it missed
    /// the WS event).
    pub passkey_pending: Arc<std::sync::Mutex<Option<String>>>,
    /// Resolves a pending passkey ceremony with the app's assertion (or cancel).
    /// `None` when no passkey ceremony is waiting.
    pub passkey_answer: Arc<std::sync::Mutex<Option<tokio::sync::oneshot::Sender<PasskeyAnswer>>>>,
}

impl AppState {
    pub async fn new(cfg: Arc<Config>) -> Result<Self> {
        let store = HistoryStore::new(cfg.clone()).await?;
        let (tx, _rx) = broadcast::channel(256);
        Ok(Self {
            cfg,
            store: Arc::new(store),
            tx,
            current_qr: Arc::new(std::sync::Mutex::new(String::new())),
            phone: Arc::new(std::sync::Mutex::new(String::new())),
            calls: crate::calls::CallRegistry::default(),
            pair_code: Arc::new(std::sync::Mutex::new(String::new())),
            passkey_pending: Arc::new(std::sync::Mutex::new(None)),
            passkey_answer: Arc::new(std::sync::Mutex::new(None)),
        })
    }

    pub fn is_linked(&self) -> bool {
        !self.phone.lock().unwrap().is_empty()
    }

    pub fn phone(&self) -> String {
        self.phone.lock().unwrap().clone()
    }

    pub fn set_qr(&self, qr: &str) {
        *self.current_qr.lock().unwrap() = qr.to_string();
    }

    pub fn set_phone(&self, phone: &str) {
        *self.phone.lock().unwrap() = phone.to_string();
    }

    /// Record a new passkey request so `/passkey/request` can serve it.
    pub fn set_passkey_pending(&self, options: String) {
        *self.passkey_pending.lock().unwrap() = Some(options);
    }

    pub fn take_passkey_pending(&self) -> Option<String> {
        self.passkey_pending.lock().unwrap().take()
    }

    /// Install the oneshot the app's `/passkey/assertion` resolves. Replaces
    /// (and cancels) any prior pending ceremony.
    pub fn set_passkey_answer(
        &self,
        tx: tokio::sync::oneshot::Sender<PasskeyAnswer>,
    ) {
        let mut slot = self.passkey_answer.lock().unwrap();
        if let Some(old) = slot.take() {
            let _ = old.send(PasskeyAnswer::Cancelled);
        }
        *slot = Some(tx);
    }

    /// Take the oneshot (so the app's answer resolves the *current* ceremony).
    pub fn take_passkey_answer(
        &self,
    ) -> Option<tokio::sync::oneshot::Sender<PasskeyAnswer>> {
        self.passkey_answer.lock().unwrap().take()
    }
}
